barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp

```diff
@@ -168,6 +186,168 @@ void Circuit<FF>::univariate_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr
+        p = (end + beg) / 2;
+
+        xor_props = get_standard_logic_circuit(p * 2, true);
+        and_props = get_standard_logic_circuit(p * 2, false);
+        CircuitSchema xor_circuit = xor_props.circuit;
+        CircuitSchema and_circuit = and_props.circuit;
+
+        xor_flag = cursor + xor_props.num_gates <= this->selectors.size();
+        and_flag = cursor + xor_props.num_gates <= this->selectors.size();
+        if (xor_flag || and_flag) {
+            for (size_t j = 0; j < xor_props.num_gates; j++) {
+                // It is possible for gates to be equal but wires to be not, but I think it's very
+                // unlikely to happen
+                xor_flag &= xor_circuit.selectors[j + xor_props.start_gate] == this->selectors[cursor + j];
+                and_flag &= and_circuit.selectors[j + and_props.start_gate] == this->selectors[cursor + j];
+
+                if (!xor_flag && !and_flag) {
+                    if (j == 0) {
+                        stop_flag = true;
+                    }
+                    break;
+                }
+            }
+        }
+        if (stop_flag) {
+            break;
+        }
+
+        if (!xor_flag && !and_flag) {
+            end = p - 1;
+        } else {
+            res = 2 * p;
+            logic_flag = xor_flag;
+
+            beg = p + 1;
+        }
+    }
+    // TODO(alex): Figure out if I need to create range constraint here too or it'll be
+    // created anyway in any circuit
+    if (res != static_cast<size_t>(-1)) {
+        xor_props = get_standard_logic_circuit(res, true);
+        and_props = get_standard_logic_circuit(res, false);
+
+        info("Logic constraint optimization: ", std::to_string(res), " bits. is_xor: ", xor_flag);
+        size_t left_gate = xor_props.gate_idxs[0];
+        uint32_t left_gate_idx = xor_props.idxs[0];
+        size_t right_gate = xor_props.gate_idxs[1];
+        uint32_t right_gate_idx = xor_props.idxs[1];
+        size_t out_gate = xor_props.gate_idxs[2];
+        uint32_t out_gate_idx = xor_props.idxs[2];
+
+        uint32_t left_idx = this->real_variable_index[this->wires_idxs[cursor + left_gate][left_gate_idx]];
+        uint32_t right_idx = this->real_variable_index[this->wires_idxs[cursor + right_gate][right_gate_idx]];
+        uint32_t out_idx = this->real_variable_index[this->wires_idxs[cursor + out_gate][out_gate_idx]];
+
+        FF left = this->symbolic_vars[left_idx];
+        FF right = this->symbolic_vars[right_idx];
+        FF out = this->symbolic_vars[out_idx];
+
+        if (logic_flag) {
+            (left ^ right) == out;
+        } else {
+            (left ^ right) == out; // TODO(alex): implement & method
+        }
+
+        optimized[left_idx] = false;
+        optimized[right_idx] = false;
+        optimized[out_idx] = false;
+        return cursor + xor_props.num_gates;
+    }
+    return res;
+}
+
+template <typename FF> size_t Circuit<FF>::range_constraint_handler(size_t cursor)
```

Comment:
Please add description. Also the name "handler" would be logical with a class. With a function you want it to be a verb form (mb handle_range_constraints or transform_range_constraints)